gradle-sls-packaging/src/main/java/com/palantir/gradle/dist/service/tasks/ModuleArgs.java

```diff
@@ -92,23 +100,40 @@ static ImmutableList<String> collectClasspathArgs(
                 .distinct()
                 .sorted()
                 .flatMap(ModuleArgs::addOpensArg);
-        return Stream.concat(exports, opens).collect(ImmutableList.toImmutableList());
-    }
+        Stream<String> enablePreview = enablePreview(javaVersion, parsedJarManifests);
 
-    private static Optional<JarManifestModuleInfo> parseModuleInfo(@Nullable java.util.jar.Manifest jarManifest) {
-        return Optional.ofNullable(jarManifest)
-                .<JarManifestModuleInfo>map(manifest -> JarManifestModuleInfo.builder()
-                        .exports(readManifestAttribute(manifest, ADD_EXPORTS_ATTRIBUTE))
-                        .opens(readManifestAttribute(manifest, ADD_OPENS_ATTRIBUTE))
-                        .build())
-                .filter(JarManifestModuleInfo::isPresent);
+        return Stream.of(exports, opens, enablePreview)
+                .flatMap(Function.identity())
+                .collect(ImmutableList.toImmutableList());
     }
 
-    private static List<String> readManifestAttribute(java.util.jar.Manifest jarManifest, String attribute) {
-        return Optional.ofNullable(
-                        Strings.emptyToNull(jarManifest.getMainAttributes().getValue(attribute)))
-                .map(ENTRY_SPLITTER::splitToList)
-                .orElseGet(ImmutableList::of);
+    private static Stream<String> enablePreview(
+            JavaVersion javaVersion, Map<File, JarManifestModuleInfo> parsedJarManifests) {
+        Map<String, Collection<String>> enablePreviewFromJar = parsedJarManifests.entrySet().stream()
+                .filter(entry -> entry.getValue().enablePreview().isPresent())
+                .collect(Multimaps.toMultimap(
+                        entry -> entry.getValue().enablePreview().get(),
+                        entry -> entry.getKey().getName(),
+                        () -> MultimapBuilder.hashKeys().arrayListValues().build()))
+                .asMap();
+
+        if (enablePreviewFromJar.size() > 1) {
+            throw new RuntimeException("Unable to add '--enable-preview' because classpath jars have embedded "
+                    + JarManifestModuleInfo.ENABLE_PREVIEW_ATTRIBUTE + " attribute with different versions:\n"
+                    + enablePreviewFromJar);
+        }
+
+        if (enablePreviewFromJar.size() == 1) {
+            String enablePreviewVersion = Iterables.getOnlyElement(enablePreviewFromJar.keySet());
+            Preconditions.checkState(
+                    enablePreviewVersion.equals(javaVersion.toString()),
+                    "Runtime java version (" + javaVersion + ") must match version from embedded "
+                            + JarManifestModuleInfo.ENABLE_PREVIEW_ATTRIBUTE + " attribute (" + enablePreviewVersion
+                            + ") from:\n" + Iterables.getOnlyElement(enablePreviewFromJar.values()));
+            return Stream.of("--enable-preview");
+        }
```

Comment:
This method is pretty complicated, I think we want to stream through the jars, fail if we see any with ENABLE_PREVIEW_ATTRIBUTE != javaVersion, and add the flag if we've seen any ENABLE_PREVIEW_ATTRIBUTE attrs?

imo the `Map<String, Collection<String>> enablePreviewFromJar` makes things difficult to grok

Reply:
Sorry for the delay, I've updated to use less stream-y stuff... think this should be a bit more readable now!